stl/inc/format

```diff
@@ -0,0 +1,2734 @@
+const _CharT* _Find_encoded(const _CharT* _First, const _CharT* _Last, const _CharT _Val, const _Cvtvec& _Cvt) {
+    // Returns the first occurrence of _Val as an encoded character (and not, for example, as a
+    // continuation byte) in [_First, _Last).
+    if (_Cvt._Mbcurmax == 1 || _Cvt._Mbcurmax == 4) {
+        // As above and in _Mbrtowc, assume 4-byte encodings are UTF-8
+        return _Find_unchecked(_First, _Last, _Val);
+    }
+
+    while (_First != _Last && *_First != _Val) {
+        const int _Units = _Code_units_in_next_character(_First, _Last, _Cvt);
+        if (_Units < 0) {
+            throw format_error("Invalid encoded character in format string.");
+        }
+        _First += _Units;
+    }
+
+    return _First;
+}
+
+template <class _CharT, _Parse_replacement_field_callbacks<_CharT> _HandlerT>
+void _Parse_format_string(basic_string_view<_CharT> _Format_str, _HandlerT&& _Handler) {
+    auto _Begin         = _Format_str.data();
+    auto _End           = _Begin + _Format_str.size();
+    const _Cvtvec& _Cvt = _Getcvt();
+
+    while (_Begin != _End) {
+        const _CharT* _OpeningCurl = _Begin;
+        if (*_Begin != '{') {
+            _OpeningCurl = _Find_encoded(_Begin, _End, _CharT{'{'}, _Cvt);
+
+            for (;;) {
+                const _CharT* _ClosingCurl = _Find_encoded(_Begin, _OpeningCurl, _CharT{'}'}, _Cvt);
+
+                // In this case there are neither closing nor opening curls in [_Begin, _OpenCurl)
+                // Write the whole thing out.
+                if (_ClosingCurl == _OpeningCurl) {
+                    _Handler._On_text(_Begin, _OpeningCurl);
+                    break;
+                }
+                // We know _ClosingCurl isn't past the end because
+                // the above condition was not met.
+                ++_ClosingCurl;
+                if (_ClosingCurl == _OpeningCurl || *_ClosingCurl != '}') {
+                    throw format_error("Unmatched '}' in format string.");
+                }
+                // We found two closing curls, so output only one of them
+                _Handler._On_text(_Begin, _ClosingCurl);
+
+                // skip over the second closing curl
+                _Begin = _ClosingCurl + 1;
+            }
+
+            // We are done, there were no replacement fields.
+            if (_OpeningCurl == _End) {
+                return;
+            }
+        }
+        // Parse the replacement field starting at _OpeningCurl and ending sometime before _End.
+        _Begin = _Parse_replacement_field(_OpeningCurl, _End, _Handler);
+    }
+}
+
+template <class _CharT>
+struct _Basic_format_specs {
+    int _Width           = 0;
+    int _Precision       = -1;
+    char _Type           = '\0';
+    _Align _Alignment    = _Align::_None;
+    _Sign _Sgn           = _Sign::_None;
+    bool _Alt            = false;
+    bool _Localized      = false;
+    bool _Leading_zero   = false;
+    uint8_t _Fill_length = 1;

```

Comment:
We *could* pack this a bit better into one `uint8_t`. Is it worth it?

Reply:
This trades time (additional bitwise operations) for space, so it isn't a pure win. I would prefer to see profiling data indicating that this is an issue, before adding code complexity here.

Reply:
if you mean combining the bools into the top bits of the `_Fill_length` I'm with STL, I don't think it's worth it. That said we could probably move things around a bit to eliminate padding bytes. Also this structure will probably be difficult to modify in the future because of ABI.

Reply:
Yeah, I think it is not worth packing it but maybe optimizing the order would be useful